app/models/concerns/dirty_associations.rb

```diff
@@ -14,12 +14,23 @@
 module DirtyAssociations
   extend ActiveSupport::Concern
 
+  included do
+    after_save :reset_state
+  end
+
   module ClassMethods
     # usage:
     #   class Model
     #     dirty_has_many_associations :organizations, :locations
     def dirty_has_many_associations(*args)
       extension = Module.new do
+        define_method "reset_state" do
+          args.each do |assoc|
```

Comment:
Given we iterate over args inside the method, this won't work if someone callse `dirty_has_many_associations` in one model several times. If you move it to [this](https://github.com/theforeman/foreman/blob/ee34c8710f5d7f3a08368ee354e0542838b32ce5/app/models/concerns/dirty_associations.rb#L34) block, you can define reset method per each arg. An example of a model that calls `dirty_has_many_associations` twice - ProvisioningTemplate. Once in the model definition itself, second time thanks to including Taxonomix concern.

Reply:
By the time after save is called args get replaced with the args of the last dirty association even if I move the reset_state code to the block. Could please clarify if my understanding is right?

Reply:
I don't think that's what happens. When `dirty_has_many_associations` is called, it only defines new methods, so you can call it many times. The method you're adding should follow the same pattern I think. I'll comment more in the code.

Reply:
Thanks for clarifying.